src/snowflake/snowpark/relational_grouped_dataframe.py

```diff
@@ -403,8 +406,36 @@ def apply_in_pandas(
             - :func:`~snowflake.snowpark.functions.pandas_udtf`
         """
 
+        partition_by = [Column(expr, _emit_ast=False) for expr in self._grouping_exprs]
+
+        # this is the case where this is being called from spark
+        # this is not handleing nested column access, it is assuming that the access in the function is not nested
+        original_columns: List[str] | None = None
+        key_columns: List[str] | None = None
+        if context._is_called_from_snowpark_connect:
+            if self._dataframe._column_map is not None:
+                original_columns = [
+                    column.spark_name for column in self._dataframe._column_map.columns
+                ]
+            signature = inspect.signature(func)
+            parameters = signature.parameters
+            if len(parameters) == 2:
+                key_columns = [
+                    unquote_if_quoted(col.get_name()) for col in partition_by
+                ]
+
         class _ApplyInPandas:
             def end_partition(self, pdf: pandas.DataFrame) -> pandas.DataFrame:
+                if key_columns is not None:
+                    import numpy as np
+
+                    key_list = [pdf[key].iloc[0] for key in key_columns]
+                    numpy_array = np.array(key_list)
+                    keys = tuple(numpy_array)
+                if original_columns is not None:
+                    pdf.columns = original_columns
+                if key_columns is not None:
+                    return func(keys, pdf)
```

Comment:
```suggestion
                if original_columns is not None:
                    pdf.columns = original_columns
                if key_columns is not None:
                    import numpy as np

                    key_list = [pdf[key].iloc[0] for key in key_columns]
                    numpy_array = np.array(key_list)
                    keys = tuple(numpy_array)
                    return func(keys, pdf)
```
nit: can we restructure it this way? grouping the if statements together

Reply:
It won't work because we are changing the pdf in 
`if original_columns is not None:
                    pdf.columns = original_columns`